extensions/resource-deployment/src/services/azdataService.ts

```diff
@@ -0,0 +1,168 @@
+'use strict';
+import * as cp from 'child_process';
+import * as path from 'path';
+import * as fs from 'fs';
+import { IPlatformService } from './platformService';
+
+export interface DeploymentProfile {
+	name: string;
+	defaultDataSize: string;
+	defaultLogSize: string;
+	master: string;
+	data: string;
+	compute: string;
+	hdfs: string;
+	nameNode: string;
+	spark: string;
+	activeDirectory: boolean;
+	hadr: string;
+	includeSpark: string;
+	gatewayPort: string;
+	appProxyPort: string;
+	sqlServerPort: string;
+	readableSecondaryPort?: string;
+	controllerPort: string;
+	defaultDataStorageClass: string;
+	defaultLogsStorageClass: string;
+	bdcJson: any;
+	controlJson: any;
+}
+
+interface BdcConfigListOutput {
+	stdout: string[];
+}
+
+export interface IAzdataService {
+	getDeploymentProfiles(): Thenable<DeploymentProfile[]>;
+}
+
+const SqlServerMasterResourceName = 'master';
+const DataResourceName = 'data-0';
+const HdfsResourceName = 'storage-0';
+const ComputeResourceName = 'compute-0';
+const NameNodeResourceName = 'nmnode-0';
+const HadrEnabledSettingName = 'hadr.enabled';
+export class AzdataService implements IAzdataService {
+	constructor(private platformService: IPlatformService) {
+	}
+
+	getDeploymentProfiles(): Thenable<DeploymentProfile[]> {
+		process.env['ACCEPT_EULA'] = 'Yes';
+		return this.ensureWorkingDirectoryExists().then(() => {
+			return this.getDeploymentProfileNames();
+		}).then((names: string[]) => {
+			const profilePromises: Thenable<DeploymentProfile>[] = [];
+			names.forEach(name => {
+				profilePromises.push(this.getDeploymentProfileInfo(name));
+			});
+			return Promise.all(profilePromises);
+		});
+	}
+
+	private getDeploymentProfileNames(): Thenable<string[]> {
+		return new Promise<string[]>((resolve, reject) => {
+			cp.exec('azdata bdc config list -o json', (error, stdout, stderror) => {
+				if (error) {
+					reject(error.message);
+				} else {
+					try {
+						const output = <BdcConfigListOutput>JSON.parse(stdout);
+						resolve(output.stdout);
+					}
+					catch (err) {
+						if (err instanceof Error) {
+							reject(err.message);
+						} else {
+							reject(err);
+						}
+					}
+				}
+			});
+		});
+	}
+
+	private getDeploymentProfileInfo(profileName: string): Thenable<DeploymentProfile> {
+		return new Promise<DeploymentProfile>((resolve, reject) => {
+			cp.exec(`azdata bdc config init --source ${profileName} --target ${profileName} --force`, { cwd: this.platformService.storagePath() }, (error, stdout, stderror) => {
+				if (error) {
+					reject(error.message);
+				} else {
+					const bdcJsonPromise = this.getJsonObjectFromFile(path.join(this.platformService.storagePath(), profileName, 'bdc.json'));
+					const controlJsonPromise = this.getJsonObjectFromFile(path.join(this.platformService.storagePath(), profileName, 'control.json'));
+					Promise.all([bdcJsonPromise, controlJsonPromise]).then((value) => {
+						const bdcJson = value[0];
+						const controlJson = value[1];
```

Comment:
In general a lot of these functions could be made a lot cleaner and use better style if you made them async and just used Promises directly. Having all these nested thens and whatnot make it difficult to follow the logic and mostly seems unnecessary.